### ISSUE TITLE
When a camera is offline, other cameras stop updating snapshots and other plugins stop responding

### ISSUE THREAD

**Issue by s1rc** [hjdhjd/homebridge-unifi-protect#114]
**Describe The Problem:**

When a camera goes offline the following happens:
1. All home app camera snapshots stop updating
2. Cameras have long delay (~45 - 60s) from when device, macOS or iOS, opens a stream in home app and when homebridge logs the stream request
3. All other homebridge plugins show "No Response" until the offline camera is connect or homebridge updates the snapshot

I tried this all on a fresh install of homebridge, with only homebridge-unifi-protect installed, the result was the same. Setting `Enable.StreamOnly.Low` has no impact on this delay when one camera is offline.

**To Reproduce:**

1. Unplug one camera, all cameras stop refreshing at 10s, eventually cameras and plugin switches show "No Response".
2. Plug camera back in and when it is ready in UniFi Protect portal, all cameras start refreshing at 10s, "No Response" clears.

**Logs:**

I timed this, there was a span of ~45 seconds between when the stream was requested in macOS or iOS and this shows in the log.
```
[9/30/2020, 14:52:30] [UniFi Protect] unifi.home.com [UDM-PRO] Driveway [UVC G4 Pro]: HomeKit video stream request received: 1280x720, 30 fps, 299 kbps.
```

**Homebridge Configuration:**

```json
{
    "controllers": [
        {
            "address": "10.0.0.1",
            "doorbellMessages": [
                {
                    "message": "BE RIGHT THERE"
                }
            ],
            "password": "***",
            "username": "***"
        }
    ],
    "options": [
        "Enable.NoiseFilter"
    ],
    "verboseFfmpeg": false,
    "platform": "UniFi Protect"
}
```

**Screenshots:**

**Environment:**

* **Homebridge Version:** 1.1.6 (used to be 1.2.3 but that no longer seems to be @latest)
* **Node Version:** 12.18.4
* **Homebridge-UniFi-Protect Plugin Version:** 3.7.7
* **Apple Device and iOS / macOS / iPadOS / tvOS Version:** MBP macOS 10.15.7, iPhone iOS 14.0.1
* **Operating System and OS Version:** Homebridge official Raspbian 1.0.4

All infrastructure is UniFi.
* UDMP: 1.8.0
* USW-Pro-24-PoE: All cameras are connected directly to switch

**Comment by hjdhjd** [comment]
Thanks for the report! How many cameras are in your environment, what are their types, and how did you execute the test? I'd like to try to recreate this in my own environment.

Much appreciated!

**Comment by s1rc** [comment]
Happy to help, love your plugin!

I have:
* 2 G4 Pro
* 1 G4 Doorbell
* 2 G3 Flex

I took a screen grab of the home app in macOS. During this time I:
1. Unplugged one of the G3 Flex cameras, in the video/gif you can see the timer counts up, no new snapshot.
2. After about 30 seconds I plug the G3 Flex back in, once it is finished initializing, the snapshot continues as normal
![out](https://user-images.githubusercontent.com/2073967/94731525-e9c23880-0332-11eb-91f5-854349f68e2c.gif)

In this video/gif the G3 Flex stream starts almost instantly, then I unplug the G4 Pro and the snapshot stops. Once that happens the long delay before homebridge registers the stream request is visible.
![out](https://user-images.githubusercontent.com/2073967/94732743-a799f680-0334-11eb-95a4-458c84752de9.gif)

**Comment by s1rc** [comment]
Something I just noticed during testing. 
1. I moved 1 G4 Pro into a different room in the home app.
2. I went back to the other room with the G3 Flex, snapshot refreshed fine, stream started fast.
3. I unplugged the G4 Pro, nothing in log.
4. The G3 Flex continued to work fine, snapshots were updating, stream was fast even after 30-60 seconds.
5. I switch to the room with the G4 Pro, homebridge logs `Unable to retrieve snapshot` after about 15 seconds.
6. I switch to the room with the G3 Flex, snapshot is no longer updated, stream will not start.
7. Wait about 5 minutes on the room with the G3 Flex, since the home app isn't requesting any snapshots for the unplugged G4 Pro in the other room, the G3 Flex starts updating and streams work normally again.

I know in Discord you mentioned that not everything in the log is to be of concern, it just seems like the unresponsive devices only happen during the time of that log message.

**Comment by hjdhjd** [comment]
I have 20 cameras in my setup, and I unplugged 5 of them and can’t replicate this issue. I’m unconvinced that the plugin is the issue here rather than something going on with Protect in your configuration. **What version of UniFi Protect are you running?**

The length of time between snapshots has very little to do with the plugin - it’s determined by HomeKit. Whether it takes 5 seconds or 500 seconds...there’s not much I can do here. :smile: So let’s set that aside for now...there’s not much to do here.

I’m sure the above looks clear to you...I can’t follow most of what you outlined above. You focus on the amount of time between actions...but it’s not really a useful metric to measure much at the moment.

_What, in the simplest of terms is broken?_ You’re trying to provide a lot of information above, but it’s not in a digestible (to me) order that I can follow.

If your underlying issue is: snapshots don’t update in a timely way in certain scenarios...there’s very little I can do to help that. I might try to streamline a couple of elements in a future release, but it’s not a material problem that I can see, and I can’t replicate a significant issue.

Oh...and stop unplugging your cameras. :smile:

**Comment by s1rc** [comment]
Sorry for all the different scenarios. In simplest terms, the issue I'm seeing is **when a camera is offline, all other cameras stop working as well as other homebridge devices**.

The snapshot time and logging was just observation, not part of the issue. Normally all cameras are online, just having recently moved I'm unplugging them to run wires. So in an ideal situation, there is no issue, but it is conceivable to have a camera go offline on its own due to PoE issues, switching issues, updates, etc. i.e. last week I upgraded my main switch FW and the cameras went offline, so all of my homebridge devices were non responsive in homekit for a while.

My UniFi Protect version is **1.14.11**, which appears to be the latest stable train release.

**Comment by hjdhjd** [comment]
Thanks for being specific...super helpful. I've duplicated the scenario you've laid out and I can't replicate the issue. In fact, I tried powering off 19 out of 20 cameras and still couldn't replicate this. I still had great responsiveness and none of the symptoms you described.

In either case: having devices plugged in or not should make no difference. **The plugin doesn't really care because it's never actually querying the cameras, it queries the Protect controller.** If you have a problem with the **controller** I can conceivably see an issue. However, short of that (which isn't the situation you described), I can't replicate your issue.

I know that you're seeing this issue, but there's something in your environment that's different / unique and that's what I'm trying to figure out. Either that, or the issue you're describing is a symptom of something else entirely...

So let's try to figure out how you're setup and what's at the heart of this.

On my end, I mentioned I'll streamline a couple of things in a future release but really, that's not the answer here. This is a problem *everyone* should have if it exists, and I haven't heard it raised...and I, who has a substantial installation myself, can't replicate. :smile:

**Comment by hjdhjd** [comment]
Give v3.7.8 a try. Enjoy.

**Comment by s1rc** [comment]
Updated this morning, unplugged two cameras and so far so good all day!! 🎉 

So weird that my setup had this issue, especially looking at your changes for the fix. Maybe something with my Raspberry Pi, since even with a clean install it still happened.

Many thanks for this and the wonderful plugin(s)! (I use your myQ also) I'll leave the cameras unplugged and keep monitoring but so far this seems resolved.

**Comment by hjdhjd** [comment]
No worries. Glad you're enjoying the plugins. :smile:

Best of luck.
